Augment DownloadFile task to make it able to download from public/private location

JohnTortugo opened the pull request:
Closes: https://github.com/dotnet/arcade/issues/4164
Related to: https://github.com/dotnet/arcade/issues/3868

Make `DownloadFile` receive a new parameter called `PrivateUri` which should contain a secondary URL to use for downloading a file in case the download from `Uri` fail.

Addressed comments from previous PR: https://github.com/dotnet/core-sdk/pull/5352

[Tested restoring from public location.]https://dnceng.visualstudio.com/internal/_build/results?buildId=406936&view=results)
[Tested restoring from _private_ location.](https://dnceng.visualstudio.com/internal/_build/results?buildId=407613&view=results)

/cc @livarcocc @mmitche @riarenas @dagood

dagood commented on src/redist/targets/GenerateLayout.targets:
```diff
@@ -288,11 +266,19 @@
+        <!--
+          Replaces the host portion of 'BaseUrl' with 'PrivateBaseUrl'.
+        -->
+        <PrivateBaseUrl Condition="'$(InternalBuild)' == 'true'">$(MSRCBaseURL)$([System.String]::new('%(BaseUrl)').Substring($([System.String]::new('%(BaseUrl)').IndexOf('/', 8))))</PrivateBaseUrl>
```
Unclear why `8` is chosen. It seems like this could just be a `Replace`.

JohnTortugo replied:
Yeah. Sorry about that. I had a newer version of this file where I actually use a Replace and I don't have the <Error.../> thing below.

dagood commented on src/redist/targets/GenerateLayout.targets:
```diff
@@ -288,11 +266,19 @@
     </ItemGroup>
-  
+
+    <Error Condition="$([System.String]::new('%(ComponentToDownload.BaseUrl)').IndexOf('http')) == -1" Text="The BaseURL '%(BaseUrl)' needs to be set with the 'http[s]' prefix." />
```
This is a bug, the task needs to accept `file://` schema.

Error message is also not accurate, this seems to checks for `http` appearing anywhere in the string.

dagood replied:
In general it seems like the logic here would be easier to properly write error-handling for inside the MSBuild task.

dagood commented on src/core-sdk-tasks/DownloadFile.cs:
```diff
@@ -14,12 +16,25 @@ public class DownloadFile : Task
         public string Uri { get; set; }
 
+        /// <summary>
+        /// If this field is set and the task fail to download the file from `Uri` it will try
```
Description is a little wrong, it only falls back to private URI if it hits a 404, not simply on failure.

dagood commented on src/core-sdk-tasks/DownloadFile.cs:
```diff
@@ -39,31 +54,57 @@ public override bool Execute()
                 File.Copy(filePath, DestinationPath);
+                return true;
             }
             else
```
`else` is unnecessary now, removing helps with the somewhat deep nesting.

JohnTortugo replied:
@dagood - I'm not sure why the else is unecessary. Shouldn't the task still support download from a `file://` location?

dagood replied:
I mean that since you added the exit with `return true;`, the `else {` and matching `}` can be removed to improve readability. (Keep what's inside, unindent it.) Not sure how that affects `file://` support.

dagood commented on src/core-sdk-tasks/DownloadFile.cs:
```diff
@@ -39,31 +54,57 @@ public override bool Execute()
-                        throw;
-                    }
+
+                    Thread.Sleep(new Random().Next(3000, 10000));
```
A few things:
* `await Task.Delay(...)` probably isn't any different for this particular PR, but would be more expected in an async method.
* Creating a new `Random` repeatedly can cause unexpected non-randomness due to seeding with current time (doesn't seem to matter in this case, but something to keep in mind).
* Instead of random, consider exponential falloff as a way to help better when the problem is resource contention. https://github.com/dotnet/arcade/blob/master/src/Common/ExponentialRetry.cs

JohnTortugo replied:
Thanks Davis. I'll adjust this.

JohnTortugo commented:
Thanks for the comments @dagood . I think I've addressed all of them. Latest test build is here: https://dnceng.visualstudio.com/internal/_build/results?buildId=408033&view=results

dagood commented on src/core-sdk-tasks/DownloadFile.cs:
```diff
@@ -39,31 +54,56 @@ public override bool Execute()
             }
 
-            return true;
+            return false;
```
Does any actionable error show up as when it fails this way? Since nothing was actually thrown I'm worried you'd have to look all the way to the binlog to diagnose even the basics.

JohnTortugo replied:
Added a `throw HttpResponseException` here.

dagood commented on src/core-sdk-tasks/DownloadFile.cs:
```diff
@@ -39,31 +54,56 @@ public override bool Execute()
+                            if (httpResponse.StatusCode == HttpStatusCode.NotFound ||
+                                httpResponse.ReasonPhrase.IndexOf("The requested URI does not represent any resource on the server.", StringComparison.OrdinalIgnoreCase) >= 0)
+                            {
+                                return false;
```
It doesn't look like it emits an error here either...

JohnTortugo replied:
Added a `throw FileNotFoundException` here. Going up in next push.

dagood commented on src/core-sdk-tasks/DownloadFile.cs:
```diff
@@ -39,31 +54,56 @@ public override bool Execute()
+                        {
+                            if (!string.IsNullOrEmpty(PrivateUri))
+                            {
+                                httpResponse = await httpClient.GetAsync(PrivateUri);
```
This flow is a little weird with retry logic... it seems to mean *both* requests have to succeed in a single retry attempt to actually go through, but ideally it should only retry the private URI once it's established the public one is a 404.

I also wonder if each one should have its own retry count, so that failure to request the public URI doesn't reduce the number of chances it has to request the private one.

That leads me to think that maybe the request/copy logic should move into a different method, and this one should have two retry loops: one tries to download from the public blob storage, the second tries to download from private storage if the first found a 404.

This also makes it easy to print out an error that's specific to the endpoint being used, rather than the unclear `'{Uri}'/'{PrivateUri}'` below.

I don't know if this will ever matter in practice, just a suggestion. Sorry I didn't catch what was happening here in earlier iterations.

JohnTortugo replied:
Done.

JohnTortugo commented:
Addressed all feedback. Test build (download part) passed: https://dnceng.visualstudio.com/internal/_build/results?buildId=411792&view=results

I'm afraid this PR is becoming a DownloadFile optimization PR... I don't think that's high priority right now and I'll postpone all non-essential changes to another PR.

dagood commented on src/core-sdk-tasks/DownloadFile.cs:
```diff
@@ -39,31 +54,70 @@ public override bool Execute()
             }
-            else
+
+            var lastStatus = await DownloadWithRetries(Uri, DestinationPath);
```
Use explicit `bool` since the type isn't obvious on the right side. I also think `success` would be a simpler name:

```suggestion
            bool success = await DownloadWithRetries(Uri, DestinationPath);
```

dagood replied:
Oh wait. Use explicit `bool?`. This was really confusing!

dagood commented on src/core-sdk-tasks/DownloadFile.cs:
```diff
@@ -39,31 +54,70 @@ public override bool Execute()
+
+            for (int retryNumber = 0; retryNumber < MaxRetries; retryNumber++)
+            {
                 using (var httpClient = new HttpClient())
```
It's actually better to use a single instance rather than create and dispose a bunch of `HttpClient`s. This doesn't seem like an extreme case but FYI. https://softwareengineering.stackexchange.com/q/330364

dagood commented on src/core-sdk-tasks/DownloadFile.cs:
```diff
@@ -39,31 +54,70 @@ public override bool Execute()
+        /// <param name="source">URL to the file to be downloaded.</param>
+        /// <param name="target">Local path where to put the downloaded file.</param>
+        /// <returns>true: Download Succeeded. false: Download failed with 404. null: Download failed but is retriable.</returns>
+        private async Task<bool?> DownloadWithRetries(string source, string target)
```
Missing async suffix:

```suggestion
        private async Task<bool?> DownloadWithRetriesAsync(string source, string target)
```

dagood commented on src/core-sdk-tasks/DownloadFile.cs:
```diff
@@ -39,31 +54,70 @@ public override bool Execute()
                         {
-                            getTask.Result.CopyTo(outStream);
+                            Log.LogMessage($"File not found error while attempting download using the following URL(s): '{source}'.");
+                            return false;
```
What happened to the plan to add diagnosability here?

JohnTortugo replied:
Done: https://dnceng.visualstudio.com/internal/_build/results?buildId=411924&view=results

dagood commented on src/core-sdk-tasks/DownloadFile.cs:
```diff
@@ -39,31 +54,75 @@ public override bool Execute()
                     {
-                        File.Delete(DestinationPath);
-                        throw;
+                        Log.LogMessage($"Problems downloading file from '{source}'. {e.Message} {e.StackTrace}");
```
This message also looks like it'll still only show up in the binlog.

JohnTortugo replied:
Changed it.

JohnTortugo commented on src/core-sdk-tasks/DownloadFile.cs:
```diff
@@ -39,31 +54,75 @@ public override bool Execute()
-                Log.LogMessage(MessageImportance.High, $"Downloading '{Uri}' to '{DestinationPath}'");
 
-                using (var httpClient = new HttpClient())
+            var downloadStatus = await DownloadWithRetriesAsync(Uri, DestinationPath);
```
Missed that :-(

JohnTortugo commented on src/core-sdk-tasks/DownloadFile.cs:
```diff
@@ -39,31 +54,75 @@ public override bool Execute()
+
+            if (downloadStatus != true)
+            {
+                throw new FileNotFoundException($"Error while attempting download file from the provided URLs.");
```
Makes sense. Made it an Exception.

JohnTortugo commented:
Comments addressed. Validated on these builds:

- [Success download the file.](https://dnceng.visualstudio.com/internal/_build/results?buildId=411924&view=results)
- [Fail to download the file](https://dnceng.visualstudio.com/internal/_build/results?buildId=412114&view=results).

JohnTortugo commented on src/core-sdk-tasks/DownloadFile.cs:
```diff
@@ -39,31 +54,75 @@ public override bool Execute()
+                        if (httpResponse.StatusCode == HttpStatusCode.NotFound ||
+                            httpResponse.ReasonPhrase.IndexOf("The requested URI does not represent any resource on the server.", StringComparison.OrdinalIgnoreCase) >= 0)
+                        {
+                            Log.LogMessage($"File not found error while attempting download using the following URL(s): '{source}'.");
```
Changed it.

JohnTortugo commented:
New test builds after latest changes:

- [Success build](https://dnceng.visualstudio.com/internal/_build/results?buildId=412756&view=results)
- [Failing build ](https://dnceng.visualstudio.com/internal/_build/results?buildId=412756&view=results)

@dagood - please see in the logs that the issue you pointed is now resolved. Thanks.

JohnTortugo commented:
I don't have permission to merge the PR.

@nguerrera @livarcocc can you please take a look?

livarcocc commented:
@mmitche good to merge this?

livarcocc commented:
@mmitche good to merge this?

livarcocc commented:
@mmitche good to merge this?

nguerrera commented:
Let's hang on until 3.0.101 is actually out the door.
